[LBSE] Composited SVG shapes are blurry when an ancestor is upscaled

obyknovenius opened the pull request:
#### d4766b667963256e41db8c72a02613067074d834
<pre>
[LBSE] Composited SVG shapes are blurry when an ancestor is upscaled
<a href="https://bugs.webkit.org/show_bug.cgi?id=275400">https://bugs.webkit.org/show_bug.cgi?id=275400</a>

Reviewed by Nikolas Zimmermann.

When calculating `rootRelativeScaleFactor`, we should accommodate
the scale factor of every ancestor.

* LayoutTests/svg/compositing/transformed-child-in-scaled-parent-expected.xml: Added.
* LayoutTests/svg/compositing/transformed-child-in-scaled-parent.xml: Added.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::updateRootRelativeScale):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::rootRelativeScaleFactor):

Canonical link: <a href="https://commits.webkit.org/289144@main">https://commits.webkit.org/289144@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e0c93d71b2088fecd678d9197ba27ae0f3cc6258

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85241 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4976 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39672 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90368 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36282 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/87330 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5065 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12952 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66271 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24088 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88287 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77423 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46548 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/3742 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31684 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35350 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74478 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32520 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91807 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12588 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/9307 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/74816 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12817 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73260 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73928 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18344 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18331 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16758 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/4595 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12531 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/17996 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12361 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15854 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14112 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -4109,9 +4109,11 @@ void GraphicsLayerCA::updateRootRelativeScale()
-    if (auto* parentLayer = parent(); parentLayer && parentLayer->hasNonIdentityChildrenTransform())
-        rootRelativeScaleFactor = std::max(rootRelativeScaleFactor, computeMaxScaleFromTransform(parentLayer->childrenTransform()));
+    if (m_parent) {
+        if (m_parent->hasNonIdentityChildrenTransform())
```
We have to be careful here and only apply the logic for LBSE layers.

obyknovenius replied:
It's already applied only to LBSE layers because we return early from `updateRootRelativeScale()` if `m_shouldUpdateRootRelativeScaleFactor` is not set:
https://github.com/WebKit/WebKit/blob/aa49cd4d27107bd400dbbffc2542ddd5e30ab8ed/Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp#L4092-L4101
And we set `m_shouldUpdateRootRelativeScaleFactor` only for SVG layers and only if LBSE is enabled:
https://github.com/WebKit/WebKit/blob/aa49cd4d27107bd400dbbffc2542ddd5e30ab8ed/Source/WebCore/rendering/RenderLayerBacking.cpp#L370-L371
For all other layers `m_rootRelativeScaleFactor` always stays equal to `1.0f`.

nikolaszimmermann replied:
Ah great, thanks for checking.

nikolaszimmermann reviewed (changes requested):
Thanks Vitaly! As discussed, we still need to ensure this only run for LBSE layers and update pixel test results (and probably add a test where this is clearly visible).

nikolaszimmermann reviewed (commented):
Code wise this is fine - but it lacks some test coverage.

You need to add a new test that is actually enabling LBSE, and just include as test case something which is blurry now (when run with LBSE on, but without your patch) and is crisp with your change applied.

nikolaszimmermann reviewed (approved):
LGTM, r=me.

obyknovenius committed:
[LBSE] Composited SVG shapes are blurry when an ancestor is upscaled
https://bugs.webkit.org/show_bug.cgi?id=275400

Reviewed by Nikolas Zimmermann.

When calculating `rootRelativeScaleFactor`, we should accommodate
the scale factor of every ancestor.

* LayoutTests/svg/compositing/transformed-child-in-scaled-parent-expected.xml: Added.
* LayoutTests/svg/compositing/transformed-child-in-scaled-parent.xml: Added.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::updateRootRelativeScale):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
(WebCore::GraphicsLayerCA::rootRelativeScaleFactor):

Canonical link: https://commits.webkit.org/289144@main
